File: serv/src/main/java/task_manager/server/services/TaskService.java
package task_manager.server.services;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Service;
import task_manager.server.views.requests.TaskJsonModel;

import java.sql.Timestamp;
import java.util.List;
import java.util.stream.Collectors;


@Service
public class TaskService {
    private final JdbcTemplate jdbcTemplate;

    @Autowired
    public TaskService(JdbcTemplate template) {
        this.jdbcTemplate = template;
    }

    public void deleteTable() {
        jdbcTemplate.execute("DROP TABLE IF EXISTS users CASCADE");
    }

    private static final RowMapper<TaskJsonModel> TASK_MAPPER = (responce, num) ->
            new TaskJsonModel(
                    responce.getInt("global_id"),

                    responce.getInt("author_id"),
                    responce.getInt("changed_by"),

                    responce.getInt("priority"),
                    responce.getString("name"),
                    responce.getString("about"),

                    responce.getTimestamp("deadline"),
                    responce.getTimestamp("notificationTime"),
                    responce.getTimestamp("lastChangeTime"),

                    responce.getBoolean("completed"),
                    responce.getBoolean("deleted")
            );

    public void updateTasks(List<TaskJsonModel> tasks) {
        for (TaskJsonModel task : tasks) {
            jdbcTemplate.update(
                    "UPDATE Tasks SET " +
                            " author_id = ?," +
                            " changed_by = ?, " +
                            " priority = ?, " +
                            " name = ?, " +
                            " about = ?, " +
                            " deadline = ?, " +
                            " notificationTime = ?, " +
                            " lastChangeTime = ?, " +
                            " completed = ?, " +
                            " deleted = ? " +
                    "WHERE global_id = ? AND lastChangeTime < ?;",
                    task.getAuthorId(),
                    task.getChangedBy(),
                    task.getPriority(),
                    task.getName(),
                    task.getAbout(),
                    task.getDeadline(),
                    task.getNotificationTime(),
                    task.getLastChangeTime(),
                    task.isCompleted(),
                    task.isDeleted(),

                    task.getGlobalId(),
                    task.getLastChangeTime()
            );
        }
    }

    public List<TaskJsonModel> createTasks(List<TaskJsonModel> tasks) {
        for (TaskJsonModel task : tasks) {
            int taskId = jdbcTemplate.queryForObject(
                    "INSERT INTO Tasks (author_id, changed_by, priority, name, about, deadline, notificationTime, lastChangeTime, completed, deleted)" +
                            " VALUES(?,?,?,?,?,?::TIMESTAMP WITH TIME ZONE,?::TIMESTAMP WITH TIME ZONE,?::TIMESTAMP WITH TIME ZONE,?,?) " +
                            " RETURNING global_id;",
                    Integer.class,
                    task.getAuthorId(),
                    task.getChangedBy(),
                    task.getPriority(),
                    task.getName(),
                    task.getAbout(),
                    task.getDeadline(),
                    task.getNotificationTime(),
                    task.getLastChangeTime(),
                    task.isCompleted(),
                    task.isDeleted()
            );
            task.setGlobalId(taskId);
        }

        return tasks;
    }



    public List<TaskJsonModel>  getUpdatedTask(Timestamp syncTime, int authorId, List<Integer> excludedTasks) {
        List<TaskJsonModel> tasks;
        if (syncTime == null) {
            tasks = jdbcTemplate.query(
                    "SELECT * FROM Tasks " +
                            "WHERE " +
                            "author_id = ? ;",
                    TASK_MAPPER,
                    authorId
            );
        } else {
            tasks = jdbcTemplate.query(
                    "SELECT * FROM Tasks " +
                            "WHERE " +
                            "lastChangeTime > ?::TIMESTAMP WITH TIME ZONE AND " +
                            "author_id = ?;",
                    TASK_MAPPER,
                    syncTime,
                    authorId
            );
        }

        return tasks.stream()
                .filter(task -> !excludedTasks.contains(task.getGlobalId()) ).collect(Collectors.toList());
    }
}

File: serv/src/main/java/task_manager/server/services/TaskService.kt
package task_manager.server.services

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.RowMapper
import task_manager.server.models.Task
import org.springframework.stereotype.Service
import java.sql.ResultSet
import javax.validation.constraints.NotNull
import java.util.ArrayList

@Service
class TaskService @Autowired
constructor(private val jdbcTemplate: JdbcTemplate) {

    fun deleteTable() {
        jdbcTemplate.execute("DROP TABLE IF EXISTS tasks CASCADE")
    }

    fun addTask(@NotNull task: Task): Int {
        return jdbcTemplate.queryForObject<Int?>(
                "INSERT INTO tasks(author_id, caption, about, checked) VALUES(?, ?, ?, ?)" + "RETURNING id",
                arrayOf(
                        task.author_id,
                        task.caption,
                        task.about,
                        task.checked
                ),
                { response: ResultSet, rowNum: Int -> response.getInt("id") }
        )!!
    }

    fun getTaskById(id: Int?): Task {
        return jdbcTemplate.queryForObject<Task>(
                "SELECT * FROM tasks WHERE (tasks.id) = ?",
                arrayOf(id)
        ) { response, rowNum ->
            Task(
                    response.getInt("id"),
                    response.getInt("author_id"),
                    response.getString("caption"),
                    response.getString("about"),
                    response.getBoolean("checked")
            )
        }!!
    }

    fun getTasksByAuthorId(author_id: Int): List<Task> {
        return jdbcTemplate.query(
                "SELECT * FROM tasks WHERE (tasks.author_id) = ?",
                arrayOf(author_id),
                TASK_MAPPER)
    }


    fun updateTask(task: Task, id: Int?): Task {
        val querry = StringBuilder()

        querry.append("UPDATE tasks SET ")
        var changed = false

        if (task.caption != null) {
            querry.append("caption = '" + task.caption + "',")
            changed = true
        }

        if (task.about != null) {
            querry.append("about = '" + task.about + "',")
            changed = true
        }

        if (task.checked != null) {
            querry.append("checked = " + (if (task.checked!!) "True" else "False") + ",")
            changed = true
        }

        querry.deleteCharAt(querry.length - 1)
        querry.append(" WHERE tasks.id = '$id';")
        if (changed) {
            jdbcTemplate.update(querry.toString())
        }
        return task
    }

    companion object {
        private val TASK_MAPPER = { responce: ResultSet, num: Int ->
            Task(
                    responce.getInt("id"),
                    responce.getInt("author_id"),
                    responce.getString("caption"),
                    responce.getString("about"),
                    responce.getBoolean("checked")
            )
        }
    }
}